subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfiguration.java

```diff
@@ -1589,6 +1623,44 @@ private ConfigurationIdentity getIdentity() {
         return new DefaultConfigurationIdentity(buildPath, projectPath, name);
     }
 
+    private boolean isProperUsage(ProperMethodUsage... properUsages) {
+        for (ProperMethodUsage properUsage : properUsages) {
+            if (properUsage.isProperUsage(this)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    private void asertValidUsage(String methodName, ProperMethodUsage... properUsages) {
+        if (!isProperUsage(properUsages)) {
```

Comment:
If these checks were being added to non-public methods, I would agree.  But these are primarily the first call made from a configuration's client, and when they aren't then they are already guarded by the pre-existing `assertIsResolvable` checks.

If seeing extra assertions in this case hasn't happened yet, I don't think it's going to be a problem.  And since we're the only ones deprecating, that should be a signal to double-check if something can be deprecated.

I'd rather leave this for now.

Reply:
It's added to `public` methods _on internal API_.  We want to fail hard there, since we should be the only ones knowing about and calling these directly.  If they are called from public API mehtods, then the existing guards on the public API should have already failed hard.  

I don't think we should change how this works at this point.

Reply:
OK, I suppose in this case it's better not failing and only warning.  

I think this isn't a great example, since `deprecateForResolution` isn't public API, so the only ones calling it should be us, and as there are no existing tests that fail, a failure would be more useful to us to immediately realize we're not acting sensibly here.  But I concede the point, I'll make calling when the usage is deprecated not fail and only warn for these internal methods.

Reply:
A test case which fails without internal APIs:

The signatures configuration is created by us, and is deprecated for resolution

```
        buildFile << """
            plugins {
                id 'signing'
            }

            ${mavenCentralRepository()}

            dependencies {
                signatures 'org.apache.commons:commons-lang3:3.9'
            }

            task resolve {
                configurations.signatures.files
            }
        """

        expect:
        succeeds("resolve", '--stacktrace')
```

> OK, I suppose in this case it's better not failing and only warning.

This method is named `asertValidUsage` (it also has a typo), and is meant for asserting. I don't think it should be warning at all, but just needs to be more restrictive on when it throws exceptions.